feat: topology (#7)
libp2p/js-interfaces: src/topology/README.md

interface-topology
========================

> Implementation of the topology interface used by the `js-libp2p` registrar.

> This interface has two main purposes. It uniforms the registration of libp2p protocols and enables a smarter connection management.

Reviewer: I think something like this would be good for now, and we can point to a more in depth overview once that's ready.

```suggestion
Topologies can be used in conjunction with `js-libp2p` to help shape its network and the overlays of its subsystems, such as pubsub and the DHT.
```

## Table of Contents

- [Implementations](#implementations)
- [Install](#install)
- [Modules using the interface](#modulesUsingTheInterface)
- [Usage](#usage)
- [Api](#api)

## Implementations

### Topology

A libp2p topology with a group of common peers.

### Multicodec Topology

A libp2p topology with a group of peers that support the same protocol.

## Install

```sh
$ npm install libp2p-interfaces
```

## Modules using the interface

TBA

## Usage

###  Topology

```js
const Topology = require('libp2p-interfaces/src/topology')

const toplogy = new Topology({
  min: 0,
  max: 50
})
```

### Multicodec Topology

```js
const MulticodecTopology = require('libp2p-interfaces/src/topology/multicodec-topology')

const toplogy = new MulticodecTopology({
  min: 0,
  max: 50,
  multicodecs: ['/echo/1.0.0'],
  handlers: {
    onConnect: (peerInfo, conn) => {},
    onDisconnect: (peerInfo) => {}
  }
})

> 

Reviewer: I would just delete these 3 lines. We can explain the registrar property in the API section.
```suggestion
```

> // Needs to set registrar in order to listen for peer changes

Reviewer: ```suggestion
```

> topology.registrar = registrar

Reviewer: ```suggestion
```

```

## API

The `MulticodecTopology` extends the `Topology`, which makes the `Topology` API a subset of the `MulticodecTopology` API.

###  Topology

- `Topology`

>   - `peers.set<function(id, PeerInfo)>`: Sets a peer in the topology.

Reviewer: Express this as a Map.
```suggestion
  - `peers<Map<string, PeerInfo>>`: A Map of peers belonging to the topology.
```

>   - `disconnect<function(PeerInfo)>`: Disconnects a peer from the topology.

Reviewer: ```suggestion
  - `disconnect<function(PeerInfo)>`: Called when a peer has been disconnected
```

#### Constructor

```js
const toplogy = new Topology({
  min: 0,
  max: 50,
  handlers: {
    onConnect: (peerInfo, conn) => {},
    onDisconnect: (peerInfo) => {}
  }
})
```

**Parameters**
- `properties` is an `Object` containing the properties of the topology.
  - `min` is a `number` with the minimum needed connections (default: 0)
  - `max` is a `number` with the maximum needed connections (default: Infinity)
  - `handlers` is an optional `Object` containing the handler called when a peer is connected or disconnected.
    - `onConnect` is a `function` called everytime a peer is connected in the topology context.
    - `onDisconnect` is a `function` called everytime a peer is disconnected in the topology context.

#### Set a peer

- `topology.peers.set(id, peerInfo)`

Add a peer to the topology.

**Parameters**

> - `id` is the `b58string` that identifies the peer to add.

Reviewer: It will eventually be base32, so maybe just leave it generic
```suggestion
- `id` is the `string` that identifies the peer to add.
```

- `peerInfo` is the [PeerInfo][peer-info] of the peer to add.

#### Notify about a peer disconnected event

- `topology.disconnect(peerInfo)`

**Parameters**
- `peerInfo` is the [PeerInfo][peer-info] of the peer disconnected.

###  Multicodec Topology

- `MulticodecTopology`

>   - `registrar<Registrar>`: Sets the registrar in the topology.

Reviewer: ```suggestion
  - `registrar<Registrar>`: The `Registrar` of the topology. This is set by the `Registrar` during registration.
```

>   - `peers.set<function(id, PeerInfo)>`: Sets a peer in the topology.

Reviewer: Let's just show that this is a Map and can be used accordingly. 
```suggestion
  - `peers<Map<string, PeerInfo>>`: The Map of peers that belong to the topology
```

  - `disconnect<function(PeerInfo)>`: Disconnects a peer from the topology.

#### Constructor

```js
const toplogy = new MulticodecTopology({
  min: 0,
  max: 50,
  multicodecs: ['/echo/1.0.0'],
  handlers: {
    onConnect: (peerInfo, conn) => {},
    onDisconnect: (peerInfo) => {}
  }
})
```

**Parameters**
- `properties` is an `Object` containing the properties of the topology.
  - `min` is a `number` with the minimum needed connections (default: 0)
  - `max` is a `number` with the maximum needed connections (default: Infinity)
  - `multicodecs` is a `Array<String>` with the multicodecs associated with the topology.
  - `handlers` is an optional `Object` containing the handler called when a peer is connected or disconnected.
    - `onConnect` is a `function` called everytime a peer is connected in the topology context.
    - `onDisconnect` is a `function` called everytime a peer is disconnected in the topology context.

> #### Set the registrar

Reviewer: Let's delete this section, I think the property overview is sufficient for now. Also, the Registrar is going to set `topology.registrar` on the base Topology as well during registration, every Topology will have that property once they've registered, so it might be good to add that to the Topology

Author: Removed the section, added the `set` to the `Topology` and overwrite it in the `MulticodecTopology`
